### ISSUE TITLE
Yapbam crash report function doesn't work anymore

### ISSUE THREAD

**Issue by fathzer** [yapbam/yapbam#9]
The web service that sends a notification in the Yapbam/crashes Teams canal returns a 502 error ... and this makes Yapbam crash again :-(
In Yapbam, we have to inform the user and invite him to send an email when this happens.

We also have to fix the 502 error in the web service
Maybe we have to find a way more robust than Teams (The "Gérer le canal" menu in Teams currently hangs!!!) to send alerts. The problem is sending email has become very complicated due to security constraints :-(

**Comment by marcoM32** [comment]
The @fathzer,
meanwhile, thank you for including me in the yapbam community for me it is a great honor. Regarding this issue, I have an idea to solve the problem by perhaps improving the usability as well.

I have been dealing with PHP to Telegram integrations very often and perhaps we could include for Yapbam this type of development creating a private channel in which to send error reports. Telegram is robust enough to hardly ever crack in the face of well-constructed inputs so it could fix the random errors causing che 502 erroron crashReport.php service. 
To do this type of evolution, we could mainly act on the YapbamWeb product (which should also be registered here on Git Hub in my opinion) and later you could think of modifying the client by mapping the data on a body json and not through simple parameters post data.
The alternative is to send the reports via email in order to have them directly available related from the application log.
What do you think, Also based on the frequency of use of this service?

In this regard, I propose the following libraries to be included via "composer" for the above developments:

- https://github.com/PHPMailer/PHPMailer
- https://github.com/TelegramBot/Api
- https://github.com/slimphp/Slim [Only if we want to turn the service into RESTful]

**Comment by fathzer** [comment]
Hi Marco,

Well... I should give you some context elements ...
I will try to explain my point of view (and hopefully make things clearer in my own head ;-)).
- There's already a [YapbamWeb repository on Sourceforge](https://sourceforge.net/p/yapbam/code/HEAD/tree/YapbamWeb). I didn't transfer it to Github because I considered it as "should be completely rewritten"!  So maybe I can transfer it to Github in a private repo (and give you the access).
The idea is: The web site was written more than 10 years ago with php, because at this time, Sourceforge (and others) offered free php hosting solutions suitable for the project needs. Unfortunately, the poor stability of those solutions, their poor ability to evolve (I had many problem trying to set up https), the limited set of php extensions available, convinced me to move to Docker and a dedicated server.  
Currently, the Yapbam web site runs in a Docker container on a Cloud server. Rarely, I still have problems (The site was closed when my visa card expired, sometime the https certificates are not renewed, etc...).
- I am a very, very, very bad php developer. Every development is a nightmare. 10 years after, it still sounds strange to me (F... I forgot to install that library on the server, F... again, it doesn't work with Japanese chars, etc...). Even debugging an existing program is a nightmare.  Probably because I never used it at work.
So, **the less php we add to the web site, the more comfortable I am** ;-)  
Nevertheless, **having a Docker image that exposes a web service to send alerts would be perfect** ... even if it uses php ... as long as I don't have to maintain it.
- I have a "maybe one day" project to move this web site to a more modern server less platform (like AWS Lambda or GCP cloud functions). I didn't study it a lot ... but I have some experience at work that makes me think it is possible. I hope it would help to fix some problems I have:
  - The small one: If I suddenly die ... I will (probably ;-)) stop paying for the server, and the web site will disappear in a month. As the traffic is very low, the free tier program of AWS/GCP will be enough to continue serving the web site.
  - The big one: PHP. As php isn't supported by those platform, it's a good reason to switch to a more modern platform like Node.js (and maybe to have other nightmares ;-)).

  So, **not too much work on the web site... maybe it's a waste of time**.

- I have a preference for email. I used it at the beginning of the project to alert me on crash. BUT I also have a long experience of security rules written by Google and others to prevent me from sending email :-( I developed a cool [Docker image to send email via GMail](https://github.com/fathzer/docker-mail-service), but it stopped working ... many times, too many times :-(  
I also tried some php libraries (not the one you mentioned) ... but I still had problems (for example, the docker component had a host name different from its DNS name, and GMail suddenly refused the mail my server was sending).  
After spending a couple of week-ends on that f...g mail problem, I switched to a less secure (more usable) solution: Teams, and promise I will not spend a second on email solutions anymore ;-).  
I'm absolutely not found of Teams, it is just an easy to implement solution I use since a couple of years at work without any problem ... until today ;-) Telegram, why not ... it's another way, as long as it is simple ...   
So, **if you succeed in having a Docker container sending email, and can explain me how to configure it to have gmail accept its mails, I will buy it** ;-)

- The problem if we change the input of to a json body is to keep compatibility with Yapbam previous version. About 20% of the (very few) Yapbam users, are using a version older than 1 year (<=0.20.13). So, I would prefer not to abandon them ;-) Maybe we could support both?

- The service is not used very often. As it only usage is to inform me when a Yapbam application crashes, my goal is to make it useless :-)

Nevertheless, after investigating, it seems the problem is not with Teams: Calling the Teams web hook URL with Postman works fine. So, one more time something is wrong with the php part ... I will continue to search ... tomorrow ;-)

Best regards,
Jean-Marc

**Comment by marcoM32** [comment]
Hi @fathzer,
I understand your point of view on the site and on the service in question. Personally, unlike you, I like php a lot and I have enough fun programming on it. If you wish I can rewrite the site or part of it but I would rewrite in this language because on the web level I do not know anything about doker or nodejs so I would find myself very difficult if I had to work on it. For this specific problem, if you wish, I try to check the crashReport.php service even in debug operation and if I find a problem I propose a possible fix or improvement to be applied to the class ok? As you say if the Teams service is working correctly it could come from the installation or by general configuration of the site but I can still try to check if I find something; alternatively I could try to look at the docker project you published [here](https://github.com/fathzer/docker-mail-service) because about this operation I have a nice idea to overcome the problem, if this project were to start working would you be able to release it on your current domain? 
Let me know if you find anything about this bug and if you like to review some aspects of the site in php standard. I also saw some other bugs arrive, tell me what is the priority of work.

**Comment by fathzer** [comment]
Hi,
I have just 1 minute to tell you about my bug chasing ;-)
It seems there's a DNS problem on yapbam server. It is not able to ping https://yapbam.webhook.office.com (unknown host) :-(
Stay tuned ;-)

**Comment by fathzer** [comment]
Hi again,

I have finally fixed the problem ; The DNS of the Docker container that powers the Yapbam site was no longer working (The Docker host server's DNS was ok). I simply restarted the container ... and the problem disappeared. IT is definitively not an exact science ;-)

I remember, 8 years ago, I was the new boss of the boss of a team that uses PHP for all their developments (about 160 small applications). I tried to move them to Java ... without any success (I presume they thought I was crazy). Finally, they moved from PHP to Node.js ... after I left ;-)  
For sure, PHP would never have had the success it had if it was as bad as I say :-)

Docker, or, more generally containerization, is, from my point of view, one of the two biggest IT innovations during these last 10 years (the other is serverless computing) ; Having almost as many little servers as you want, isolated from each other, on a single machine ... It changed my IT life ;-)

I stopped the "docker-mail-service" project because it uses a gmail account. One day, Google decided (probably to impose their own mail clients) that other applications would require a special authorization from the mail account owner. The problem is this *special authorization* expires automatically after a certain period making this solution not suitable for production :-( So, my advice is to spend no time on this subject.

To come back to your question, I think rewriting the whole site is not a priority (it's a "maybe one day" project). I suggest, if you agree, that you start to work on [improving html export of statements](https://github.com/yapbam/yapbam/issues/7).

I will continue to work on this issue to prevent multiple error dialogs when the first error dialog crashes.

Best regards,
Jean-Marc
PS: I have created a project dashboard [here](https://github.com/yapbam/yapbam/projects/1). Feel free to move tasks or add ideas in the *Candidate* column.
